### ISSUE TITLE
Lint idea: `old-functionality-hidden-behind-new-default-feature`

### ISSUE THREAD

**Issue by JSorngard** [obi1kenobi/cargo-semver-checks#914]
### Describe your use case

If a crate adds a new default feature that gates functionality that was previously always available, it would be a breaking change for a user of the crate that has `default-features = false` in the import in `Cargo.toml`.

E.g. the crate exports the function `foo` in version 0.5.0 and then adds a feature `bar` that is enabled by default in version 0.5.1, and gates the function `foo` behind it. A user of version 0.5.0 that run without default features would then encounter a breaking change if they run `cargo update`.

### Describe the solution you'd like

Check if the public API changes between the baseline and current versions when default features are disabled.

### Additional Context

I do not know a lot about how features and `#[cfg]` interact with this project, so what I propose may well be out of scope. Feel free to close this issue if so.  
I accidentally did this in one of my crates once, so unfortunately I know it can happen!

**Comment by obi1kenobi** [comment]
Good news! This is already possible today, via a suitable (and not necessarily obvious) application of the existing CLI flags.

In cargo-semver-checks, we always compare a "baseline" and a "current" version. Normally, the "baseline" is "whatever is currently on crates.io" and the "current" version is "whatever we're about to `cargo publish`".

In the use case you describe, the baseline and current are both the same source code, but with different features. We can do this by a combination of the `--only-explicit-features`, `--baseline-features`, and `--current-features` flags.

For example, assuming you're in the directory of the project you want to check
```
cargo semver-checks \
    --baseline-root .                 # the current directory is the baseline too
    --only-explicit-features          # disable the feature-selection heuristic
    --baseline-features std           # enable only the "std" feature in the baseline version
    --current-features std,default    # enable the "std" and "default" features in the current version
```

I *believe* naming the default feature as `default` should work. If it doesn't, that's almost certainly a bug and not intentional — please let us know and we'll fix it.

Please let me know if this sounds good, or if there's a use case that isn't covered by this! I'm always happy to make the tool even more useful whenever we can.

**Comment by JSorngard** [comment]
Ah, thank you for the response! I edited my comment to be clearer what comparison I want to do, but if I have understood those CLI arguments correctly the command I want is

```bash
cargo semver-checks --only-explicit-features --baseline-features baz --current-features baz
```
where the current version has added a default feature called `bar` that hides the function `foo`. However, doing this leads to an error that says that `cargo-semver-checks` failed to run `cargo update` because the crate does not have the feature `baz` (even though it does).
The error message gives me a command that is supposed to reproduce the error, but that command completes successfully.

**Comment by obi1kenobi** [comment]
Ah, thanks, there's definitely a bug here. Apparently our CLI-feature-parsing code doesn't support comma-separated feature names. Thanks for flagging the problem, and sorry for the inconvenience! I'll open a separate issue for that.

The command you listed will compare the crate in the current working directory (with only feature `baz`) against the largest-versioned release on crates.io for that crate that is `<=` the version in the working directory, again with only feature `baz`. This is a legal command, but it doesn't sound quite like what we were chatting about, I think.

Running the following command on the `semver-checks-features` branch catches the breaking change:
```console
$ cargo semver-checks --baseline-rev main --only-explicit-features --baseline-features 24bits --baseline-features 50bits --baseline-features libm --current-features 24bits --current-features 50bits --current-features libm
     Cloning main
     Parsing lambert_w v0.5.7 (current)
      Parsed [   0.763s] (current)
     Parsing lambert_w v0.5.6 (baseline)
      Parsed [   1.185s] (baseline)
    Checking lambert_w v0.5.6 -> v0.5.7 (minor change)
     Checked [   0.008s] 82 checks: 81 pass, 1 fail, 0 warn, 7 skip

--- failure function_missing: pub fn removed or renamed ---

Description:
A publicly-visible function cannot be imported by its prior path. A `pub use` may have been removed, or the function itself may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.35.0/src/lints/function_missing.ron

Failed in:
  function lambert_w::sp_lambert_w0, previously in file /tmp/lambert_w/target/semver-checks/git-main/a09016532578dced9c723272f6d65d3d26fc47da/src/lib.rs:137

     Summary semver requires new major version: 1 major and 0 minor checks failed
    Finished [   1.968s] lambert_w
```

Here I used a baseline of "the local repo's `main` branch", a current version of "the current directory (implicitly)", and enabled the `libm,24bits,50bits` features in both, albeit a bit verbosely to avoid comma-separated lists.

Running with `--baseline-root . --baseline-features default` works as well, and produces the same outcome. It uses the current directory as baseline, and enables the default features. This simulates the difference between "the current version on `default` features" vs "the current version with the previous-version's-default features enabled explicitly."

**Comment by JSorngard** [comment]
Thank you so much for your help! I'll see about adding some semver-checks commands to my CI then :D

**Comment by obi1kenobi** [comment]
Happy to help! Thanks for helping us find a bug as well!
